blog/_posts/2019-01-18-fluxdiffeq.md

```diff
@@ -0,0 +1,823 @@
+a problem, this will be the most efficient way to do adjoint sensitivity
+analysis. And this method only applies to ODEs. Not only that, it doesn't even
+apply to all ODEs. For example, ODEs with discontinuities ([events](http://docs.juliadiffeq.org/latest/features/callback_functions.html)) are excluded by the assumptions of the derivation.
+Thus once again we arrive at the conclusion that one method is not enough.
+
+In DifferentialEquations.jl we have implemented many different methods for
+computing the derivatives of differential equations with respect to parameters.
+We have a [recent preprint](https://arxiv.org/abs/1812.01892) detailing
+some of these results. One of the things we have found is that direct use of
+automatic differentiation can be one of the most efficient and flexible methods.
+Julia's ForwardDiff.jl, Flux.jl, and ReverseDiff.jl can directly be applied to
+perform automatic differentiation on the native Julia differential equation
+solvers themselves, and this can increase performance while giving new features.
+Our findings show that forward-mode automatic differentiation is fastest when
+there are less than 100 parameters in the differential equations, and that for
+a large enough number of parameters adjoint sensitivity analysis is the most efficient. Even
+then, we have good reason to believe that
+[the next generation reverse-mode automatic differentiation via source-to-source AD, Zygote.jl](https://julialang.org/blog/2018/12/ml-language-compiler),
+will be more efficient than all of the adjoint sensitivity implementations for
+large numbers of parameters.
+
+It is worth noting that using automatic differentiation in this form is almost
+impossible for libraries which utilize precompiled shared libraries that call
+C/C++/Fortran packages (SciPy, deSolve), and so this is a unique feature of the
+Julia differential equation suite. While entire ODE solver libraries could in
+theory be recoded into PyTorch or TensorFlow, doing so for a fully featured
+suite would be a difficult process without much promise for performance. For
+example, it is well-known that production-quality ODE solvers in
+C++/Fortran/Julia make heavy use of mutation in first class nonlinear functions
+to stave off the heavy memory pressure of allocating small arrays in a loop,
+but this kind of workflow is
+[mentioned in the documentation as not well-optimized in PyTorch](https://pytorch.org/docs/master/notes/autograd.html#in-place-operations-with-autograd). Other concerns of this fashion
+make Julia an exceptional choice for this application.
+
+Altogether, being able to switch between different gradient methods without changing
+the rest of your code is crucial for having a scalable, optimized, and
+maintainable framework for integrating differential equations and neural networks.
+And this is precisely what FluxDiffEq.jl gives the user direct access to. There
+are three functions:

```

Comment:
`There are three functions:` => `There are three functions with a similar API:`

And remove `With a similar API.` below.